### ISSUE TITLE
support shared-memory parallelism (multithreading)

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#1790]
I realize that this would require major effort, but I'm hoping that you will keep this on your radar screen: the lack of native support for shared-memory parallelism is a huge shortcoming of Julia.

The reason to support shared-memory parallelism is that it is usually far, far easier to program than distributed-memory parallelism.   Especially if you adopt the same infrastructure (already in gcc) as OpenMP or Cilk+, in which the programmer indicates which function calls and loops can be executed in parallel and the runtime system decides how many threads to spawn and how to distribute the load (e.g. by work stealing to parallelize irregular problems, as in Cilk).

Julia is attractive because it combines ease of programming with performance near that of C, but the latter is no longer true if you can simply put "#pragma parallel for" in your C loop and get 10x speedup on your 12-core machine.  Distributed arrays cannot compete with this in ease-of-use, especially for problems in which extensive inter-processor communication is required.

(Using the OpenMP runtime system is also nice in that this way the runtime system will share the same worker threads with C libraries like OpenBLAS and FFTW, so that Julia and external C code won't fight over the CPUs.)

**Comment by stevengj** [comment]
I think you mean the `threading` branch?  I looked for the `threads` branch and accidentally created a branch of this name (and promptly deleted it) since no such branch exists.

**Comment by ViralBShah** [comment]
There used to be a threads branch, and a newer threading branch, and eventually the threads branch was removed, I believe. I believe @JeffBezanson and @kpamnany are getting this up to speed this week. Perhaps worth connecting with them.

**Comment by JeffBezanson** [comment]
We got a fair amount accomplished this week. Here's the status.

- The latest branch is `jn/threading`, and it is up to date with master.
- Threading works with the new GC. Still using a barrier, but @vtjnash has some good progress towards stopping threads with a signal.
- Progress on signal handling by @vtjnash .
- Several more things in the runtime are thread safe.
- @kpamnany has started work on I/O. I believe the plan is to use a single I/O thread.
- We have a `JULIA_ENABLE_THREADING` flag to allow inclusion of the threading code, but disabled.

The plan is to merge into master soon after 0.4 branches, with `JULIA_ENABLE_THREADING` off by default. We need to start making sure everything builds and works in this state on all platforms. We want to build and also run as much of the thread-related code as possible even when threads are disabled.

**Comment by staticfloat** [comment]
I'm sensing an impending Travis build matrix addition.  :)

I can hear the server fans spinning up from here.  Good thing that Travis is roflscale.

**Comment by jakebolewski** [comment]
We will soon get dedicated hardware for you to play with.  Let the build matrix grow with impunity.

**Comment by ViralBShah** [comment]
Should we close this in favour of more specific issues dealing with multi-threading at this point?

**Comment by tkelman** [comment]
You said that last october. Judging by

> I don't think this issue should be closed until threading is merged in some form. You can track an issue much more easily than you can track a branch.

I think the answer there is "not until something gets merged."

**Comment by timholy** [comment]
Very exciting to hear about the progress here.

I can tell you're just trying to keep me following master rather than focusing on 0.4. Your devious strategy just might work.

**Comment by JeffBezanson** [comment]
The `jn/threading` branch doesn't seem to build at the moment. I get

```
/home/jeff/src/julia/src/codegen.cpp: In function ‘void jl_extern_c(jl_function_t*, jl_value_t*, jl_value_t*, char*)’:
/home/jeff/src/julia/src/codegen.cpp:1108:90: error: no matching function for call to ‘llvm::GlobalAlias::create(llvm::Type*, unsigned int, llvm::GlobalValue::LinkageTypes, char*&, llvm::Function*&, llvm::Module*)’
                             GlobalValue::ExternalLinkage, name, llvmf, llvmf->getParent());
```

and several other errors. @vtjnash Know anything about this? Am I doing something wrong?

**Comment by kpamnany** [comment]
@Keno says in https://github.com/JuliaLang/julia/pull/13410: "Just needs a rebase of the threading patches on top of llvm svn."

I fixed the compile errors by removing the ``#ifdef LLVM38`` and letting the ``#ifdef LLVM37`` branch be compiled, wherever there were errors.

**Comment by vtjnash** [comment]
it sounds like someone needs to rebase the patches on top of the llvm 3.7.0 release so that it can remain stable and buildable.

**Comment by JeffBezanson** [comment]
I'm using the `jn/tlsrebaseagainagain` branch of JuliaLang LLVM. Why would that stop working? Maybe I'm missing a step in trying to clean my build state?

**Comment by tkelman** [comment]
Something got merged, so close? It's not enabled by default yet but seems to work at the moment. #9336 covers the prereq LLVM version, and other separate issues can be filed for the occasional bug or segfault that will get found in using the threading code.

**Comment by StefanKarpinski** [comment]
It seems to me that this will be really fun to close when we do finally actually have a non-experimental threading interface.

**Comment by vtjnash** [comment]
We'll continue making progress at thread-safety with PRs like #17250, but I think the original issue is resolved and I don't see any sub-issues mentioned later (which should be opened as separate issues anyways).

**Comment by tkelman** [comment]
The programming model of what you can do with `@threads` is still pretty limited, but that can be discussed and developed in future issues/PR's.
